File: v2/src/path.ts
import { GuidId } from './gid/id'
import type { Id } from './gid/id'
import type { Gid } from './gid/gid'
import type { Maybe } from './maybe'

export type Path = { rootSlot: GuidId, edges: GuidId[] }

export function rootPath(rootSlot: GuidId): Path { return { rootSlot, edges: [] } }

export function childPath(parent: Path, label: GuidId): Path {
  return { rootSlot: parent.rootSlot, edges: [...parent.edges, label] }
}

export function popPath(path: Path): Maybe<{ parent: Path, label: GuidId }> {
  return path.edges.length === 0
    ? undefined
    : { parent: { rootSlot: path.rootSlot, edges: path.edges.slice(0, -1) }, label: path.edges[path.edges.length - 1] }
}

export function pathsEqual(a: Path, b: Path): boolean {
  return a.rootSlot.equals(b.rootSlot) && a.edges.length === b.edges.length && a.edges.every((id, i) => id.equals(b.edges[i]))
}

export function isRootPath(path: Path): boolean { return path.edges.length === 0 }

export function pathNode(gid: Gid, rootNode: Maybe<Id>, path: Path): Maybe<Id> {
  return path.edges.reduce<Maybe<Id>>(
    (node, label) => node instanceof GuidId ? gid(node)?.get(label) : undefined,
    rootNode
  )
}

File: v2-rust/src/path.rs
use crate::id::{GuidId, Id};
use crate::mutgid::MutGid;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Path {
    pub root_slot: GuidId,
    pub edges: Vec<GuidId>,
}

impl Path {
    pub fn root(root_slot: GuidId) -> Self {
        Self {
            root_slot,
            edges: Vec::new(),
        }
    }

    pub fn child(&self, label: GuidId) -> Self {
        let mut edges = self.edges.clone();
        edges.push(label);
        Self {
            root_slot: self.root_slot.clone(),
            edges,
        }
    }

    pub fn pop(&self) -> Option<(Path, GuidId)> {
        if self.edges.is_empty() {
            None
        } else {
            let parent = Path {
                root_slot: self.root_slot.clone(),
                edges: self.edges[..self.edges.len() - 1].to_vec(),
            };
            let label = self.edges[self.edges.len() - 1].clone();
            Some((parent, label))
        }
    }

    pub fn is_root(&self) -> bool {
        self.edges.is_empty()
    }

    pub fn node<'a>(&self, gid: &'a MutGid, root_node: Option<&'a Id>) -> Option<&'a Id> {
        let mut current = root_node?;
        for label in &self.edges {
            let guid = current.as_guid()?;
            current = gid.get(guid, label)?;
        }
        Some(current)
    }
}
